# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/gl/GLXLibrary.h
# Commit: c1bd0996764c
# Full Hash: c1bd0996764c49174c9169fe5550905ce5dcef88
# Author: Jed Davis <jld@mozilla.com>
# Date: 2021-07-06 21:42:42
# Regressor Bug: 1635451
# File Overlap Count: 1
# Description:
#   Bug 1635451 - Allow GLX to work in headless content processes. r=jgilbert
#   
#   This follows what we're already doing for EGL: a refcounted object
#   which can own the X connection, where we hold a weak reference from
#   the library object so that multiple contexts opportunistically share
# ==============================================================================

diff -r 2974b1ba2beb -r c1bd0996764c gfx/gl/GLXLibrary.h
--- a/gfx/gl/GLXLibrary.h	Tue Jul 06 07:42:41 2021 +0000
+++ b/gfx/gl/GLXLibrary.h	Tue Jul 06 07:42:42 2021 +0000
@@ -8,6 +8,8 @@
 
 #include "GLContextTypes.h"
 #include "mozilla/Assertions.h"
+#include "mozilla/DataMutex.h"
+#include "mozilla/gfx/XlibDisplay.h"
 #include "prlink.h"
 typedef realGLboolean GLboolean;
 
@@ -33,172 +35,174 @@
 
 class GLXLibrary final {
  public:
-  bool EnsureInitialized();
+  bool EnsureInitialized(Display* aDisplay);
 
  private:
   class WrapperScope final {
     const GLXLibrary& mGlx;
     const char* const mFuncName;
+    Display* const mDisplay;
 
    public:
-    WrapperScope(const GLXLibrary& glx, const char* const funcName);
+    WrapperScope(const GLXLibrary& glx, const char* const funcName,
+                 Display* aDisplay);
     ~WrapperScope();
   };
 
  public:
 #ifdef DEBUG
-#  define DECL_WRAPPER_SCOPE const WrapperScope wrapperScope(*this, __func__);
+#  define DECL_WRAPPER_SCOPE(display) \
+    const WrapperScope wrapperScope(*this, __func__, display);
 #else
-#  define DECL_WRAPPER_SCOPE
+#  define DECL_WRAPPER_SCOPE(display)
 #endif
 
   void fDestroyContext(Display* display, GLXContext context) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fDestroyContext(display, context);
   }
 
   Bool fMakeCurrent(Display* display, GLXDrawable drawable,
                     GLXContext context) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fMakeCurrent(display, drawable, context);
   }
 
   XVisualInfo* fGetConfig(Display* display, XVisualInfo* info, int attrib,
                           int* value) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fGetConfig(display, info, attrib, value);
   }
 
   GLXContext fGetCurrentContext() const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(nullptr)
     return mSymbols.fGetCurrentContext();
   }
 
   GLXFBConfig* fChooseFBConfig(Display* display, int screen,
                                const int* attrib_list, int* nelements) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fChooseFBConfig(display, screen, attrib_list, nelements);
   }
 
   XVisualInfo* fChooseVisual(Display* display, int screen,
                              int* attrib_list) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fChooseVisual(display, screen, attrib_list);
   }
 
   GLXFBConfig* fGetFBConfigs(Display* display, int screen,
                              int* nelements) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fGetFBConfigs(display, screen, nelements);
   }
 
   GLXContext fCreateNewContext(Display* display, GLXFBConfig config,
                                int render_type, GLXContext share_list,
                                Bool direct) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fCreateNewContext(display, config, render_type, share_list,
                                       direct);
   }
 
   int fGetFBConfigAttrib(Display* display, GLXFBConfig config, int attribute,
                          int* value) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fGetFBConfigAttrib(display, config, attribute, value);
   }
 
   void fSwapBuffers(Display* display, GLXDrawable drawable) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fSwapBuffers(display, drawable);
   }
 
   const char* fQueryExtensionsString(Display* display, int screen) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fQueryExtensionsString(display, screen);
   }
 
   const char* fGetClientString(Display* display, int screen) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fGetClientString(display, screen);
   }
 
   const char* fQueryServerString(Display* display, int screen, int name) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fQueryServerString(display, screen, name);
   }
 
   GLXPixmap fCreatePixmap(Display* display, GLXFBConfig config, Pixmap pixmap,
                           const int* attrib_list) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fCreatePixmap(display, config, pixmap, attrib_list);
   }
 
   GLXPixmap fCreateGLXPixmapWithConfig(Display* display, GLXFBConfig config,
                                        Pixmap pixmap) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fCreateGLXPixmapWithConfig(display, config, pixmap);
   }
 
   void fDestroyPixmap(Display* display, GLXPixmap pixmap) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fDestroyPixmap(display, pixmap);
   }
 
   Bool fQueryVersion(Display* display, int* major, int* minor) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fQueryVersion(display, major, minor);
   }
 
   void fBindTexImage(Display* display, GLXDrawable drawable, int buffer,
                      const int* attrib_list) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fBindTexImageEXT(display, drawable, buffer, attrib_list);
   }
 
   void fReleaseTexImage(Display* display, GLXDrawable drawable,
                         int buffer) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fReleaseTexImageEXT(display, drawable, buffer);
   }
 
   void fWaitGL() const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(nullptr)
     return mSymbols.fWaitGL();
   }
 
   void fWaitX() const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(nullptr)
     return mSymbols.fWaitX();
   }
 
   GLXContext fCreateContextAttribs(Display* display, GLXFBConfig config,
                                    GLXContext share_list, Bool direct,
                                    const int* attrib_list) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(display)
     return mSymbols.fCreateContextAttribsARB(display, config, share_list,
                                              direct, attrib_list);
   }
 
   int fGetVideoSync(unsigned int* count) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(nullptr)
     return mSymbols.fGetVideoSyncSGI(count);
   }
 
   int fWaitVideoSync(int divisor, int remainder, unsigned int* count) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(nullptr)
     return mSymbols.fWaitVideoSyncSGI(divisor, remainder, count);
   }
 
   void fSwapInterval(Display* dpy, GLXDrawable drawable, int interval) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(dpy)
     return mSymbols.fSwapIntervalEXT(dpy, drawable, interval);
   }
 
   int fQueryDrawable(Display* dpy, GLXDrawable drawable, int attribute,
                      unsigned int* value) const {
-    DECL_WRAPPER_SCOPE
+    DECL_WRAPPER_SCOPE(dpy)
     return mSymbols.fQueryDrawable(dpy, drawable, attribute, value);
   }
-
 #undef DECL_WRAPPER_SCOPE
 
   ////
@@ -216,7 +220,7 @@
   bool HasVideoMemoryPurge() { return mHasVideoMemoryPurge; }
   bool HasCreateContextAttribs() { return mHasCreateContextAttribs; }
   bool SupportsTextureFromPixmap(gfxASurface* aSurface);
-  bool SupportsVideoSync();
+  bool SupportsVideoSync(Display* aDisplay);
   bool SupportsSwapControl() const { return bool(mSymbols.fSwapIntervalEXT); }
   bool SupportsBufferAge() const {
     MOZ_ASSERT(mInitialized);
@@ -227,6 +231,8 @@
 
   auto GetGetProcAddress() const { return mSymbols.fGetProcAddress; }
 
+  std::shared_ptr<gfx::XlibDisplay> GetDisplay();
+
  private:
   struct {
     void(GLAPIENTRY* fDestroyContext)(Display*, GLXContext);
@@ -276,6 +282,8 @@
   bool mIsNVIDIA = false;
   bool mClientIsMesa = false;
   PRLibrary* mOGLLibrary = nullptr;
+  StaticDataMutex<std::weak_ptr<gfx::XlibDisplay>> mOwnDisplay{
+      "GLXLibrary::mOwnDisplay"};
 };
 
 // a global GLXLibrary instance